impl/spanner/directory/directory.go

```diff
@@ -0,0 +1,221 @@
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rtx := client.Single()
+	defer rtx.Close()
+
+	ret := []*directory.Directory{}
+	err = rtx.Read(ctx, table, spanner.AllKeys(), directoriesCols).Do(
+		func(r *spanner.Row) error {
+			d, err := readRow(r)
+			if err != nil {
+				return err
+			}
+			if d.Deleted && !showDeleted {
+				return nil
+			}
+			ret = append(ret, d)
+			return nil
+		})
+	return ret, err
+}
+
+// Write creates a new Directory.
+func (t *Table) Write(ctx context.Context, dir *directory.Directory) error {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return err
+	}
+	// Prepare data.
+	anyPB, err := ptypes.MarshalAny(dir.VRFPriv)
+	if err != nil {
+		return err
+	}
+	anyData, err := proto.Marshal(anyPB)
+	if err != nil {
+		return err
+	}
+
+	tmap, err := proto.Marshal(dir.Map)
+	if err != nil {
+		return err
+	}
+	tlog, err := proto.Marshal(dir.Log)
+	if err != nil {
+		return err
+	}
+
+	m, err := spanner.InsertStruct(table, spannerDirectory{
+		DirectoryID:   dir.DirectoryID,
+		Map:           tmap,
+		Log:           tlog,
+		VRFPublicKey:  dir.VRF.GetDer(),
+		VRFPrivateKey: anyData,
+		MinInterval:   dir.MinInterval.Nanoseconds(),
+		MaxInterval:   dir.MaxInterval.Nanoseconds(),
+		Deleted:       false,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Apply(ctx, []*spanner.Mutation{m})
+	return err
+}
+
+// Read retriaves a directory from storage. Returns status.NotFound if the row is deleted.
+func (t *Table) Read(ctx context.Context, directoryID string, showDeleted bool) (*directory.Directory, error) {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rtx := client.Single()
+	defer rtx.Close()
+	row, err := rtx.ReadRow(ctx, table, spanner.Key{directoryID}, directoriesCols)
+	if err != nil {
+		return nil, err
+	}
+	dir, err := readRow(row)
+	if err != nil {
+		return nil, err
+	}
+	if !showDeleted && dir.Deleted {
+		return nil, status.Errorf(codes.NotFound, "Directory %v is deleted", directoryID)
+	}
+	return dir, nil
+}
+
+// unwrapAnyProto returns the proto object seralized inside a serialized any.Any
+func unwrapAnyProto(anyData []byte) (proto.Message, error) {
+	var anyPB any.Any
+	if err := proto.Unmarshal(anyData, &anyPB); err != nil {
+		return nil, err
+	}
+	var privKey ptypes.DynamicAny
+	if err := ptypes.UnmarshalAny(&anyPB, &privKey); err != nil {
+		return nil, err
+	}
+	return privKey.Message, nil
+}
+
+// SetDelete deletes or undeletes a directory.
+func (t *Table) SetDelete(ctx context.Context, directoryID string, isDeleted bool) error {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return err
+	}
+	_, err = client.Apply(ctx, []*spanner.Mutation{
+		spanner.Update(table, []string{"DirectoryID", "Deleted", "DeleteTime"},
+			[]interface{}{directoryID, isDeleted, time.Now()}),
+	})
+	return err
+}
+
+// Delete permanently deletes a directory.
+func (t *Table) Delete(ctx context.Context, directoryID string) error {
```

Comment:
Do you expect this to delete just the metadata of the directory, or the entirety of data associated with it. In the latter case it would fail dramatically, as we learned recently.

Reply:
All the data too -- yes, this is a lesson to incorporate in to V2.

Reply:
Please add a TODO above the `client.Apply` call saying that this is a bulky delete that may take a long time and fail.